Summary main stats page for Dashboards

eltercero opened the pull request:
Closes #3044

## :v: What does this PR do?

It populates the summary page with the calculated data from the csv. 

This PR includes the updated code from #3053, so it may seem a lot of changes until that PR is merged.

## :mag: How should this be manually tested?

It can be checked at: https://burjassot.gobify.net/dashboards/contratos/resumen

## :eyes: Screenshots

![Captura de pantalla 2020-05-12 a las 17 11 42](https://user-images.githubusercontent.com/545235/81711222-b3fbe080-9473-11ea-91f7-b38df372356d.png)

## :shipit: Does this PR changes any configuration file?

- [ ] new environment variable in `.env.example`?
- [ ] new entry in `config/application.yml`?
- [ ] new entry in `config/secrets.yml`?

(Changes in these files might need to update the role in Ansible)

## :book: Does this PR require updating the documentation?

- [ ] new site configuration variable?
- [ ] new site template?
- [ ] new module/submodule settings?
- [ ] significant changes in some feature?

eltercero committed:
Summary page for Gobierto Dashboards contracts

ferblape commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -1,3 +1,215 @@
+    const lessThan1000Total = contractsData.filter((contract) => parseFloat(contract.final_amount) < 1000).length;
+    const lessThan1000Pct = lessThan1000Total/numberContracts;
+
+    const largerContractAmount = d3.max(contractsData, function(contract) { return parseFloat(contract.final_amount)} );
```
`d3.max(contractsData, (contract) => parseFloat(contract.final_amount))`

ferblape commented on app/javascript/gobierto_dashboards/webapp/lib/utils.js:
```diff
@@ -0,0 +1,10 @@
+export function formatCurrency(amount) {
```
No need to define a new function here, there's a filter for currencies in `app/javascript/lib/shared/modules/vue-filters.js`. Filters can be used in the HTML of a component like these {{ amount | money }}

ferblape commented on app/javascript/gobierto_dashboards/webapp/containers/tender/TendersShow.vue:
```diff
@@ -0,0 +1,66 @@
+  methods: {
+    initFormattedTender(){
+      this.formattedTender = Object.assign({}, this.tender);
+      this.formattedTender.initial_amount = formatCurrency(this.formattedTender.initial_amount);
```
See my comment about the money filter.

ferblape commented on app/javascript/gobierto_dashboards/webapp/containers/tender/TendersShow.vue:
```diff
@@ -0,0 +1,66 @@
+      tendersData: this.$root.$data.tendersData,
+      tender: null,
+      formattedTender: null,
+      labelTenderAmount: I18n.t('gobierto_dashboards.dashboards.contracts.tender_amount'),
```
There's a filter also for translations, called `translate`, in `app/javascript/lib/shared/modules/vue-filters.js`

Instead of declaring data variables for translations I'd prefer to use it in the template like this:

`{{ 'gobierto_dashboards.dashboards.contracts.tender_amount' | translate }}`

eltercero replied:
That does not seems to work, though. I think `translate` is supposed to receive an Object and not a string?

ferblape replied:
Hmmm, correct. I want to create at some point a filter to translate strings in the template and avoid so many variables in data (it happens too in Gobierto Data). We'll addresss this in the future.

ferblape commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -1,3 +1,215 @@
+      currency: 'EUR'
+    });
 
+    document.getElementById("mean-savings").innerText = meanSavings.toLocaleString(I18n.locale, {
```
to format this amount can we use also the money filter as plain function

ferblape reviewed (commented):
I've added some details we can discuss later

eltercero committed:
Several changes after PR review

- Using money function/filter in contracts_controller and vue components
- Removing previously created currency function
- No need to use formattedContract or formattedTender either
- Changing named function to arrow function

Crashillo commented:
I'll review this once #3053 be merged

Crashillo commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -77,6 +80,10 @@ export class ContractsController {
+
+        EventBus.$on('summary_ready', () => {
+          _self._renderSummary(remoteCsvData);
+        });
```
No self/this.

```suggestion
        EventBus.$on('summary_ready', () => this._renderSummary(remoteCsvData) )
```

Crashillo commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -115,4 +122,76 @@ export class ContractsController {
+
+  _renderTendersMetricsBox(tendersData){
+    // Calculations
+    const amountsArray = tendersData.map((tender) => parseFloat(tender.initial_amount) );
```
The point of destructuring if quite useful in this sort of situations:
```suggestion
    const amountsArray = tendersData.map(({ initial_amount = 0 }) => parseFloat(initial_amount) );
```

Unless you know (because you set it up before) that a variable can't come `undefined`, this approach is better to handle data that we don't know (i.e. comes from DB). If `tender` comes null or undefined, `parseFloat` won't never fail

eltercero replied:
This is great, thanks!

Crashillo commented on app/javascript/gobierto_dashboards/modules/contracts_controller.js:
```diff
@@ -115,4 +122,76 @@ export class ContractsController {
+    const amountsArray = tendersData.map((tender) => parseFloat(tender.initial_amount) );
+
+    const numberTenders = tendersData.length;
+    const sumTenders = d3.sum(amountsArray);
+    const meanTenders = d3.mean(amountsArray);
+    const medianTenders = d3.median(amountsArray);
```
`import { sum, mean, median }` from 'd3-array'

![imagen](https://user-images.githubusercontent.com/817526/82189454-b2a53a80-98ef-11ea-821e-01e2aa4f1deb.png)

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/contract/ContractsShow.vue:
```diff
@@ -1,18 +1,18 @@
 <template>
   <div>
-    <h1>{{ formattedContract.title }}</h1>
+    <h1>{{ contract.title }}</h1>
```
I notice you don't use `contract` object in any way. What if you declare each variable (title, description, ...) in the data() and only work with its properties?

eltercero replied:
Will do!

eltercero replied:
@Crashillo now I have this:

```javascript 
    this.title = contract.title
    this.description = contract.description
    this.assignee = contract.assignee
    this.document_number = contract.document_number
    this.final_amount = contract.final_amount
    this.initial_amount = contract.initial_amount
    this.status = contract.status
    this.process_type = contract.process_type
```

Is there a way to do it with destructuring?

Crashillo replied:
Two steps:
```
    const {
      title,
      description,
      assignee,
      document_number,
      final_amount,
      status,
      process_type
    } = contract

this.title = title
....
```

One step (perhaps more confusing)
```
    ({
      title: this.title,
      description: this.description,
      assignee: this.assignee,
      document_number: this.document_number,
      final_amount: this.final_amount,
      initial_amount: this.initial_amount
      status: this.status,
      process_type: this.process_type
    } = contract)
```

You might want to initialize them also.

eltercero replied:
Thanks!

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/summary/Summary.vue:
```diff
@@ -1,13 +1,113 @@
+    this.labelMeanSavings = I18n.t('gobierto_dashboards.dashboards.contracts.summary.mean_savings'),
+    this.labelLessThan1000 = I18n.t('gobierto_dashboards.dashboards.contracts.summary.label_less_than_1000'),
+    this.labelLargerContractAmount = I18n.t('gobierto_dashboards.dashboards.contracts.summary.label_larger_contract_amount'),
+    this.labelHalfSpendingsContracts = I18n.t('gobierto_dashboards.dashboards.contracts.summary.label_half_spendings_contracts')
```
To data()

eltercero replied:
Sorry, I forgot about this one!

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/tender/TendersShow.vue:
```diff
@@ -1,9 +1,9 @@
 <template>
   <div>
-    <h1>{{ formattedTender.title }}</h1>
+    <h1>{{ tender.title }}</h1>
```
Some concept as Contracts. Is actually so different this component from the contracts one that we need another component?

eltercero replied:
Tenders page may not be in the final versión. If it finally is, I'll merge the logic.

Crashillo commented on config/locales/gobierto_dashboards/views/ca.yml:
```diff
@@ -6,18 +6,34 @@ ca:
+        summary:
+          contracts: Contractes adjudicats
+          contracts_for: contractes per import de
+          label_half_spendings_contracts: El <strong> <span id = "half-spendings-contracts-pct">
+            </ span> </ strong> de contractes concentren el 50% de tota la despesa
+          label_larger_contract_amount: El major contracte suposa un <strong> <span
+            id = "larger-contract-amount-pct"> </ span> </ strong> de tota la despesa
+            en contractes
+          label_less_than_1000: El <strong><span id="less-than-1000-pct"></span></strong>
```
HTML in a translation block?

eltercero replied:
Yeah, I thought about this one for a while. I've just realized that in Budget there's a suffix "_html" in the key for this kind of translations.  Is there a better way to approach this problem?

Crashillo replied:
I've seen the file. This way is quite a lot ofuscated, as you're adding an `id` to the block which you're gonna update that value later with a `document.getElementById`, is that right?
This is barely legible.
I would split the translations into smaller parts, to leave here just plain text

eltercero replied:
So, spliting in smaller parts would make these translation keys. There are quotation marks because otherwise it'd be easy to miss the required whitespaces (although it can also be fixed through js).

```yaml 
label_half_spendings_contracts_1: "El "
label_half_spendings_contracts_2: " de contratos concentran el 50% de todo el gasto"
```

Later in vue I'd need to do this (triming just in case). 

```javascript
let arrayHalfSpendingsContracts = [
  I18n.t('gobierto_dashboards.dashboards.contracts.summary.label_half_spendings_contracts_1'),
  '<strong><span id="half-spendings-contracts-pct"></span></strong>',
  I18n.t('gobierto_dashboards.dashboards.contracts.summary.label_half_spendings_contracts_2')
]

this.labelHalfSpendingsContracts = arrayHalfSpendingsContracts.map((str) => str.trim()).join(' ')
```

What do you think? All in all, I'm not entirely sure this adds more clarity.

Crashillo replied:
No, the goal is avoid messing with the logic/stuff. HTML in the template tag, javascript in the script tag, translation in the yml, and so on.
So this array is also mixing unalike stuff

eltercero committed:
No need to use _self.

eltercero committed:
Some fixes post PR review

- Moving label assignation to the data initialization.
- Importing d3 in contracts_controller
- Using destructuring with array functions
- Spliting contract variable into parts since we don't really use the main object

eltercero committed:
Using destructuring.

eltercero committed:
Fix calculations and tests

eltercero commented:
@ferblape @Crashillo I'm merging this PR if that's ok, there are two other PRs (#3063, #3075) that are depending on this one.

ferblape commented:
Ok, to me yes
